### DISCUSSION TITLE
TypeError: Cannot read property 'pathname' of null (production system down)

### DISCUSSION THREAD

**Discussion by omar-dulaimi** [vercel/next.js#15557] [Help] [answered]
Hello,
The production system is down after a recent deployment yesterday, although the same code works at the dev website.

Could anybody please tell me what might the issue be here? 
I couldn't find anything online.

UPDATE: commenting out this part from _document does actually solve the problem, but why?

```
<Link href={`/search`} prefetch={false}>
                <a
                    className="d-inline-block text-center"
                    style={{ textDecoration: 'none', color: darkColor }}
                >
                    <Search
                        width="23px"
                        height="23px"
                        color={darkColor}
                        style={{ display: 'block', margin: '0 auto' }}
                    />
                    <div className="icon-lbl">
                        <span>search</span>
                    </div>
                </a>
            </Link>
```

```

web-app_1  | TypeError: Cannot read property 'pathname' of null
web-app_1  |     at Link (webpack-internal:///./node_modules/next/dist/client/link.js:159:14)
web-app_1  |     at processChild (/usr/src/app/node_modules/react-dom/cjs/react-dom-server.node.development.js:3043:14)
web-app_1  |     at resolve (/usr/src/app/node_modules/react-dom/cjs/react-dom-server.node.development.js:2960:5)
web-app_1  |     at ReactDOMServerRenderer.render (/usr/src/app/node_modules/react-dom/cjs/react-dom-server.node.development.js:3435:22)
web-app_1  |     at ReactDOMServerRenderer.read (/usr/src/app/node_modules/react-dom/cjs/react-dom-server.node.development.js:3373:29)
web-app_1  |     at renderToStaticMarkup (/usr/src/app/node_modules/react-dom/cjs/react-dom-server.node.development.js:4004:27)
web-app_1  |     at renderDocument (/usr/src/app/node_modules/next/dist/next-server/server/render.js:3:594)
web-app_1  |     at renderToHTML (/usr/src/app/node_modules/next/dist/next-server/server/render.js:47:72)
web-app_1  |     at runMicrotasks (<anonymous>)
web-app_1  |     at processTicksAndRejections (internal/process/task_queues.js:97:5)
web-app_1  |     at async /usr/src/app/node_modules/next/dist/next-server/server/next-server.js:71:285
web-app_1  |     at async /usr/src/app/node_modules/next/dist/next-server/server/next-server.js:70:142
web-app_1  |     at async DevServer.renderToHTMLWithComponents (/usr/src/app/node_modules/next/dist/next-server/server/next-server.js:92:300)
web-app_1  |     at async DevServer.renderToHTML (/usr/src/app/node_modules/next/dist/next-server/server/next-server.js:93:255)
web-app_1  |     at async DevServer.renderToHTML (/usr/src/app/node_modules/next/dist/server/next-dev-server.js:30:578)
web-app_1  |     at async DevServer.render (/usr/src/app/node_modules/next/dist/next-server/server/next-server.js:50:236)
web-app_1  |     at async Object.fn (/usr/src/app/node_modules/next/dist/next-server/server/next-server.js:35:306)
web-app_1  |     at async Router.execute (/usr/src/app/node_modules/next/dist/next-server/server/router.js:38:67)
web-app_1  |     at async DevServer.run (/usr/src/app/node_modules/next/dist/next-server/server/next-server.js:44:494)
web-app_1  |     at async DevServer.handleRequest (/usr/src/app/node_modules/next/dist/next-server/server/next-server.js:13:101)
```

**Comment by zestzero** [comment]
same here, after updating next to `9.5.0`

**Reply by zestzero** [reply]
Just found that there was a fix already https://github.com/vercel/next.js/commit/86ba29f65468ae822c561bf1d8a5ae679c58a0dc

**Reply by cvazquez** [reply]
I just ran into the same problem waking up this morning. I thought I made a change yesterday that broke it and have been re-installing and reverting code. How is the fix applied?

**Comment by ijjk** [comment, marked as answer]
Hi, this should be resolved in `v9.5.1-canary.1` of Next.js specifically this PR https://github.com/vercel/next.js/pull/15604

**Reply by omar-dulaimi** [reply]
Hello @ijjk 
We use next v9.2.1 and we're affected by this issue. Why does that happen?  Do we have to upgrade now to fix it?

**Reply by ijjk** [reply]
@omar-dulaimi this specific error is from [a change](https://github.com/vercel/next.js/pull/14804) introduced in `v9.4.5-canary.29` of Next.js, if you are seeing this in `v9.2.1` then it is caused by something else. I would recommend upgrading to the latest stable version though `v9.5.1`

**Reply by omar-dulaimi** [reply]
Thank you @ijjk

**Reply by omar-dulaimi** [reply]
@ijjk I just found out the problem. I use npm for development, and yarn in Dockerfile for deployment. So apparently yarn was getting the latest versions by itself, ignoring package.json

Changing all yarn instances in Dockerfile to npm, like npm install and npm run build instead of yarn and yarn build, solved the problem.

What gave it away was the server logs, when the server starts it loads env vars from .env and .env productions which is a feature introduced in later versions than v9.2.1

Anyway, glad it's fixed for other people too. Thank you!

**Comment by evantahler** [comment]
I'm seeing a regression, and this bug is back with next.js v11 + React 17

**Reply by ghost** [reply]
Yes I can also confirm this came back

**Reply by dy-byron** [reply]
Same here
For now, I could patch it doing something like this to the whole instances in the project, but it feels bad:
```
const { locale, push } = useRouter() || { locale: '', push: () => {} }
```
